sdk/src/main/scala/io/horizen/account/state/ForgerStakeStorageV3.scala

```diff
@@ -0,0 +1,404 @@
+    val stakeHistory = StakeHistory(stakeKey)
+    val stakeHistorySize = stakeHistory.getSize(view)
+    stakeHistory.updateOrAddCheckpoint(view, stakeHistorySize, epochNumber, latestStake => latestStake.add(stakedAmount))
+    if (stakeHistorySize == 0)
+      addNewDelegator(view, forgerKey, delegatorPublicKey)
+  }
+
+  def removeStake(view: BaseAccountStateView,
+                  blockSignProposition: PublicKey25519Proposition,
+                  vrfPublicKey: VrfPublicKey,
+                  epochNumber: Int,
+                  delegatorPublicKey: Address,
+                  stakedAmount: BigInteger): Unit = {
+
+
+    val forgerKey = getForgerKey(blockSignProposition, vrfPublicKey)
+    val forgerHistory = StakeHistory(forgerKey)
+    val forgerHistorySize = forgerHistory.getSize(view)
+    if (forgerHistorySize == 0)
+      throw new ExecutionRevertedException("Forger doesn't exist.")
+
+    val stakeKey = getStakeKey(forgerKey, delegatorPublicKey)
+    val stakeHistory = StakeHistory(stakeKey)
+    val stakeHistorySize = stakeHistory.getSize(view)
+    if (stakeHistorySize == 0)
+      throw new ExecutionRevertedException(s"Delegator ${BytesUtils.toHexString(delegatorPublicKey.toBytes)} doesn't have stake with the forger.")
+
+
+    def subtractStake(latestStake: BigInteger): BigInteger = {
+      val newAmount = latestStake.subtract(stakedAmount)
+      if (newAmount.signum() == -1)
+        throw new ExecutionRevertedException(s"Not enough stake. Claimed $stakedAmount, available $latestStake")
+      newAmount
+    }
+
+    stakeHistory.updateOrAddCheckpoint(view, stakeHistorySize, epochNumber, subtractStake)
+    forgerHistory.updateOrAddCheckpoint(view, forgerHistorySize, epochNumber, subtractStake)
+  }
+
+  def getAllForgerStakes(view: BaseAccountStateView): Seq[(PublicKey25519Proposition, VrfPublicKey, Address, BigInteger)] = {
+    val listOfForgerKeys = ForgerMap.getKeys(view)
+    listOfForgerKeys.flatMap { forgerKey =>
+      val forger = ForgerMap(view, forgerKey)
+      val delegatorList = DelegatorList(forgerKey)
+      val delegatorSize = delegatorList.getSize(view)
+      val listOfStakes: ListBuffer[(PublicKey25519Proposition, VrfPublicKey, Address, BigInteger)] = ListBuffer()
+      for (idx <- 0 until delegatorSize) {
+        val delegator = delegatorList.getDelegatorAt(view, idx).address()
+        val stakeKey = getStakeKey(forgerKey, delegator)
+        val stakeHistory = StakeHistory(stakeKey)
+        val amount = stakeHistory.getLatestAmount(view)
+        if (amount.signum() == 1)
+          listOfStakes.append((forger.forgerPublicKeys.blockSignPublicKey, forger.forgerPublicKeys.vrfPublicKey, delegator, amount))
+      }
+      listOfStakes
+    }
+
+
+  }
+
+
+  case class StakeHistory(uid: Array[Byte])
+    extends StateDbArray(ACCOUNT, Blake2b256.hash(Bytes.concat(uid, "History".getBytes("UTF-8")))) {
+
+    def addCheckpoint(view: BaseAccountStateView, epoch: Int, stakeAmount: BigInteger): Unit = {
```

Comment:
I would add an additional check that the epoch in input is > last epoch we have in the storage.
otherwise throw an exception

Reply:
I have added the check in updateOrAddCheckpoint, because it is the only place where a checkpoint can be added to an existing stake history.